### ISSUE TITLE
nyc+jest reports unknown for unit tests, works fine for integration tests

### ISSUE THREAD

**Issue by mnmkng** [istanbuljs/nyc#729]
Hello,

I was trying to get NYC (latest version) running with no success the past afternoon. We're using Jest (latest version) as our test runner and unfortunately, running Jest with the `--coverage` flag didn't work for us in integration tests (getting "unknowns"), so I added NYC to our deployment scripts.

Interestingly, it works perfect in our integration tests that spawn a server using beforeEach() and correctly outputs coverage data. However, it doesn't work at all in any (unit) test, where a subprocess server is not spawned. If a project contains some tests that spawn a server and some that don't, coverage will be reported for the ones that spawn a server and missed for all other.

Jest `--coverage` works for our unit tests and correctly reports coverage, so we will temporarily use Jest to cover unit tests and NYC to cover integration tests. A single solution would be greatly appreciated though.

I've provided a minimal repo that reproduces the issue. There's really nothing more than NYC and Jest, a trivial function and a trivial test, so it seems the problem might lie somewhere in the combination of NYC & Jest?

Thank you for any feedback / ideas.

## Expected Behavior
NYC should correctly report coverage after running unit tests with Jest.

## Observed Behavior
The reports below are taken from runs against the provided repo.

NYC reports unknown coverage:
```
Ran all test suites.
----------|----------|----------|----------|----------|----------------|
File      |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------|----------|----------|----------|----------|----------------|
All files |  Unknown |  Unknown |  Unknown |  Unknown |                |
----------|----------|----------|----------|----------|----------------|
```

When ran with the --all flag, the report is still broken:

```
Ran all test suites.
----------|----------|----------|----------|----------|----------------|
File      |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------|----------|----------|----------|----------|----------------|
All files |        0 |      100 |        0 |        0 |                |
 app.js   |        0 |      100 |        0 |        0 |          1,3,6 |
----------|----------|----------|----------|----------|----------------|
```
Interestingly, running jest with --coverage flag works fine:
```
Ran all test suites.
----------|----------|----------|----------|----------|----------------|
File      |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------|----------|----------|----------|----------|----------------|
All files |    66.67 |      100 |       50 |    66.67 |                |
 app.js   |    66.67 |      100 |       50 |    66.67 |              6 |
----------|----------|----------|----------|----------|----------------|

```
## Bonus Points! Code (or Repository) that Reproduces Issue
https://github.com/mnmkng/nyc-jest-issue

## Forensic Information

**Operating System:** MacOS High Sierra 10.13.1
**Environment Information:** Node 8.9.0, NPM 5.5.1.

**Comment by JaKXz** [comment]
`jest` has `istanbuljs` libraries built in, so you do not need to set it up with `nyc`.

**Comment by mnmkng** [comment]
@JaKXz I am aware that `jest` has Istanbul built in. It doesn't explain the issue though, nor does it help me get correct coverage.

The issue is that full coverage works neither using `nyc` nor `jest`. Either I only get unit tests working (`jest`) or integration tests (`nyc`). I did not raise the issue with `jest`, since it is mainly a testing framework and "coverage of subprocesses" seems like a niche.

`nyc` claims that it can cover subprocesses (which is why I've used it in the first place) and it does. It should also cover simple unit tests, but it doesn't. And I feel that must be a `nyc` issue?

Or perhaps `jest` does something to the files that prevents `nyc` from correctly instrumenting them?

Could you please take another quick look at it?
Thank you.

**Comment by JaKXz** [comment]
https://github.com/mnmkng/nyc-jest-issue is working as expected, as you saw: 
> Interestingly, running jest with --coverage flag works fine:

and I concur, `jest --coverage` should give you everything you need, without nyc at all.

If you have a specific issue with your tests executing code that runs in a subprocess, I would love to see an example repo for that.

**Comment by mnmkng** [comment]
@JaKXz I am sorry. I wanted to make an absolutely minimal repo focusing on the `nyc` issue and by doing that I've apparently obfuscated everything.

I've added a subprocess to the test code. I'm also committing the coverage results. As you can see, running `jest --coverage` outputs correct coverage for `app.js`, but not for `child.js` and vice versa, `nyc` correctly covers `child.js`, but doesn't touch `app.js`.

I've also run the tasks both w/ & w/o `--runInBand`. Makes no difference.

This is a report from running `npm test`:
```
Ran all test suites.
----------|----------|----------|----------|----------|----------------|
File      |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------|----------|----------|----------|----------|----------------|
All files |      100 |      100 |      100 |      100 |                |
 child.js |      100 |      100 |      100 |      100 |                |
----------|----------|----------|----------|----------|----------------|

```

And this is a report from running `npm run jest`:
```
Ran all test suites.
----------|----------|----------|----------|----------|----------------|
File      |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------|----------|----------|----------|----------|----------------|
All files |       80 |      100 |    66.67 |       80 |                |
 app.js   |       80 |      100 |    66.67 |       80 |              6 |
----------|----------|----------|----------|----------|----------------|

```

Thanks again for looking into it and sorry for not being clear right from the start.

**Comment by JaKXz** [comment]
I added this to the `package.json`:

```json
  "jest": {
    "collectCoverageFrom": [
      "src/**/*.js"
    ]
  },
```

and `child.js` shows up in the coverage report:

```
$ jest --coverage --env=node --runInBand
 PASS  test/app.test.js
  ✓ test app (2ms)
  ✓ test parent/child (86ms)

Test Suites: 1 passed, 1 total
Tests:       2 passed, 2 total
Snapshots:   0 total
Time:        0.263s, estimated 1s
Ran all test suites.
----------|----------|----------|----------|----------|----------------|
File      |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------|----------|----------|----------|----------|----------------|
All files |    57.14 |      100 |       50 |    57.14 |                |
 app.js   |       80 |      100 |    66.67 |       80 |              6 |
 child.js |        0 |      100 |        0 |        0 |            2,5 |
----------|----------|----------|----------|----------|----------------|
```

**Comment by JaKXz** [comment]
@mnmkng I think the issue here is that your `child.js` file is not exporting anything that is `import`ed or `require`d in a test file, which is why it wasn't picked up automatically by jest, and why it shows no coverage in jest's reporter.

To combine coverage reports from nyc and jest, you could try outputting the `json` coverage report from jest in the `.nyc_output` directory, and then run your tests again through nyc + jest where you just `include` the files that create sub-processes to see if nyc's coverage combining logic handles things for you. cc @bcoe, any ideas?

**Comment by bcoe** [comment]
@JaKXz @mnmkng nothing immediately comes to mind; I believe Jest still uses many of the same core Istanbul libraries that also power nyc -- so I'm not at all surprised that running the two in conjunction results in stepping on each other's toes (if I had to guess, I'd bet JavaScript files are being instrumented twice).

I like @JaKXz's recommendation; I do something similar with `nyc` itself, where we run one set of tests with `tap` and another set of tests with `mocha` -- is it possible to output the `__coverage__` variable with Jest prior to exiting, rather than outputting test reports?

@mnmkng perhaps open an issue on Jest too, and we can try to coordinate.

**Comment by mnmkng** [comment]
@JaKXz @bcoe Hello again gentlemen, I'm sorry for the long absence, "stuff happened".

I've opened the suggested issue with Jest @ https://github.com/facebook/jest/issues/5274

I've updated and retested everything and nothing has changed in the meantime. Except colleagues who run on Windows now get zero coverage from either Jest or NYC. I'll try to find out more if I can.

Thanks again for your time.

**Comment by peterdanis** [comment]
I had the same issue, but solved it by:
- instructing `jest` to output `json` files
- both `nyc` and `jest` to use the same report directory (in my case the `jest`'s default `./coverage` dir).

In `package.json`:

```
  "jest": {
    "coverageReporters": ["json"]
  },
  "nyc": {
    "reporter": ["lcov", "text"],
    "tempDirectory": "./coverage"
  }
```

Then run: `yarn nyc --clean jest --coverage`  ( or `npm run nyc --clean jest --coverage`, or without `yarn`/`npm run` if you installed it globally).

The `--clean` is important, will delete the `coverage` dir (or whatever dir you set as `tempDirectory`) before each run. This is **needed** as `nyc` will store its partial coverage files here and the file names are unique identifiers (changing each run). If you dont clean them up, **you can get false coverage** results for files, which were not tested in the last testing round.

Upper table is `jest` output, lower table is merged `jest` + `nyc` output:
![image](https://user-images.githubusercontent.com/26599181/37034771-a8649fb2-214a-11e8-9a0a-f40a3f1a59be.png)

**Comment by mnmkng** [comment]
@peterdanis 
Thank's for your solution Peter. It did not work for us, but apparently put us on the right track. We finally figured that the problem was the way in which we terminated the child processes.

Using `childProcess.kill`, at the end of our tests, no coverage would be output.

So we changed the termination procedure to `process.send` a message to the child process and have it exit itself and voila, coverage showed up.

I will close the issue now, since the problem is fixed. Thanks to all for the ideas.
